lib/benchmark/bench.go

```diff
@@ -0,0 +1,225 @@
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		start := time.Now()
+		for {
+			select {
+			case <-ticker.C:
+				delay = delay + interval
+				t := start.Add(delay)
+				measure := benchMeasure{
+					ResponseStart: t,
+					command:       c.Command,
+					client:        tc,
+					ctx:           ctx,
+					interactive:   c.Interactive,
+				}
+				requestsC <- &measure
+			case <-ctx.Done():
+				close(requestsC)
+				return
+			}
+		}
+	}()
+
+	go run(ctx, requestsC, resultC)
+
+	var result Result
+	result.ResponseHistogram = hdrhistogram.New(MinValue, MaxValue, SignificantFigures)
+	result.ServiceHistogram = hdrhistogram.New(MinValue, MaxValue, SignificantFigures)
+	results := make([]*benchMeasure, 0, c.MinimumMeasurements)
+	statusTicker := time.NewTicker(1 * time.Second)
+	timeElapsed := false
+	start := time.Now()
+	for {
+		if c.MinimumWindow <= time.Since(start) {
+			timeElapsed = true
+		}
+		select {
+
+		case measure := <-resultC:
+			results = append(results, measure)
+			if timeElapsed && len(results) >= c.MinimumMeasurements {
+				go cancel()
+			}
+			if measure.Error != nil {
+				result.RequestsFailed++
+				result.LastError = measure.Error
+			}
+			result.ResponseHistogram.RecordValue(int64(measure.End.Sub(measure.ResponseStart) / time.Millisecond))
+			result.ServiceHistogram.RecordValue(int64(measure.End.Sub(measure.ServiceStart) / time.Millisecond))
+			result.RequestsOriginated++
+		case <-ctx.Done():
+			result.Duration = time.Since(start)
+			return result, nil
+		case <-statusTicker.C:
+			log.Printf("working... current observation count: %d", len(results))
+		}
+
+	}
+}
+
+type benchMeasure struct {
+	ResponseStart time.Time
+	ServiceStart  time.Time
+	End           time.Time
+	Error         error
+	ctx           context.Context
+	client        *client.TeleportClient
+	command       []string
+	interactive   bool
+}
+
+func run(ctx context.Context, request <-chan *benchMeasure, done chan<- *benchMeasure) {
+	defer func() {
+		if r := recover(); r != nil {
+			logrus.Warningf("recover from panic: %v", r)
+		}
+	}()
+
+	for {
+		select {
+
+		case m, ok := <-request:
+			if !ok {
+				return
+			}
+			go worker(ctx, m, done)
+		case <-ctx.Done():
+			return
+		}
+
+	}
+
+}
+
+func worker(ctx context.Context, m *benchMeasure, send chan<- *benchMeasure) {
+	m.ServiceStart = time.Now()
+	execute(m)
+	m.End = time.Now()
+	select {
+	case send <- m:
+	case <-ctx.Done():
+		return
+	}
+
+}
+
+func execute(m *benchMeasure) {
+	if !m.interactive {
+		// do not use parent context that will cancel in flight requests
+		// because we give test some time to gracefully wrap up
+		// the in-flight connections to avoid extra errors
+
+		m.Error = m.client.SSH(context.TODO(), nil, false)
+		m.End = time.Now()
+		return
+	}
+	config := m.client.Config
+	client, err := client.NewClient(&config)
+	reader, writer := io.Pipe()
+	client.Stdin = reader
+	out := &bytes.Buffer{}
```

Comment:
let's use `utils.SyncBuffer` to avoid data races on concurrent writes to `Stdout` and `Stderr`

Reply:
I can't find docs for `utils.SyncBuffer`, did you happen to mean a different type of buffer? Alternatively, we can use bytes.Buffer, but make it concurrent safe and add a mutex